teps/0040-capture-step-exit-code.md

```diff
@@ -0,0 +1,70 @@
+---
+status: proposed
+title: 'Capture Step Exit Code'
+creation-date: '2021-01-06'
+last-updated: '2021-01-06'
+authors:
+- '@pritidesai'
+- '@afrittoli'
+---
+
+# TEP-0040: Capture Step Exit Code
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+  - [Use Cases](#use-cases)
+- [References](#references)
+<!-- /toc -->
+
+## Summary
+
+Tekton tasks are defined as a collection of steps in which each step can specify a container image to run.
+Steps are executed in order in which they are specified. One single step failure results in a task failure
+in turn results in a pipeline failure. Once a step results in failure, rest of the steps are not executed.
+
+Many common pipelines have requirement where a step failure must not halt the entire execution. In order to
```

Comment:
im a bit confused about whether this problem exists at the pipeline level or at the step level, is it possible to dig into it a bit more?

Reply:
The failure originates in a step and propagates to the entire pipeline. 
In terms of granularity, the proposal of this TEP is to trap the exit code for specific steps.

Reply:
I feel like we have two separate scenarios, maybe we want to handle both, but I'm not convinced both apply to the use cases we're discussing:

1. I want to stop the failure of a Task in a Pipeline from stopping the execution of a Pipeline (and maybe stop it from failing as well)
2. I want to stop the failure of a step in a Task from failing the Task

> The failure originates in a step and propagates to the entire pipeline.

Right, there is a relationship between the two - you could potentially use (2) to give you (1) as well, but (2) doesn't let you have any control over this at the pipeline level, it has to be baked into the Task. I'd argue that to truly satisfy (1) we'd want that control at the pipeline level

Reply:
I think we might eventually want support for ignoring both "step" and "task" failure.